File: uikit/src/commonMain/kotlin/com/begoml/bridge/uikit/glass/ScrollEdgeFade.kt
package com.begoml.bridge.uikit.glass

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxScope
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.Dp
import com.begoml.bridge.uikit.theme.BridgeColors

private const val PeakAlpha = 0.88f

/**
 * A dark translucent band at one end of scrolling content, with a soft edge.
 *
 * It dims what runs under it and nothing more: the content itself is never blurred. An earlier
 * version faded a blur through Haze's progressive mask, which softened the faces of the top row
 * rather than the boundary of the band.
 *
 * The ramp is a many-stop gradient on an eased curve. Density sits against the edge and the tail
 * runs long and thin, so there is no distance at which the band reads as having a border, and the
 * extra stops keep an 8-bit ramp from holding one value long enough to show as a step.
 */
@Composable
fun BoxScope.ScrollEdgeFade(
    edge: ScrollEdge,
    height: Dp,
    modifier: Modifier = Modifier,
) {
    val alignment = when (edge) {
        ScrollEdge.Top -> Alignment.TopCenter
        ScrollEdge.Bottom -> Alignment.BottomCenter
    }
    val stops = fadeStops(edge)

    Box(
        modifier = modifier
            .align(alignment)
            .fillMaxWidth()
            .height(height)
            .background(Brush.verticalGradient(colorStops = stops)),
    )
}

private const val StopCount = 12

private fun fadeStops(edge: ScrollEdge): Array<Pair<Float, Color>> = Array(StopCount) { index ->
    val t = index.toFloat() / (StopCount - 1)
    val distanceFromEdge = if (edge == ScrollEdge.Top) t else 1f - t
    val alpha = PeakAlpha * easeOut(1f - distanceFromEdge)
    t to BridgeColors.Ground.copy(alpha = alpha)
}

/** Holds density near the edge, then falls away — a shadow rather than a linear ramp. */
private fun easeOut(value: Float): Float = value * value * (3f - 2f * value)

File: iosApp/iosApp/DesignSystem/Components/ScrollEdgeFade.swift
import SwiftUI

/// A dark band at the end of the scroll, on an eased ramp.
///
/// It dims what runs under the bar and nothing more — the same job, and the same 0.88 peak, as the
/// Compose fade. The ramp is many-stopped because a short gradient on an 8-bit ramp holds one
/// value long enough to read as a step.
struct ScrollEdgeFade: View {

    enum Edge { case top, bottom }

    var edge: Edge = .bottom
    var height: CGFloat = 110

    var body: some View {
        LinearGradient(
            stops: (0..<12).map { index in
                let t = Double(index) / 11
                let distanceFromEdge = edge == .top ? t : 1 - t
                let ramp = 1 - distanceFromEdge
                let eased = ramp * ramp * (3 - 2 * ramp)
                return .init(color: Color.ground.opacity(0.88 * eased), location: t)
            },
            startPoint: .top,
            endPoint: .bottom
        )
        .frame(height: height)
        .allowsHitTesting(false)
        .ignoresSafeArea(edges: edge == .top ? .top : .bottom)
    }
}
